### PR TITLE
NDR Feature Update

### LINKED ISSUE #270: NDR threshold is automatically calculated
Users cannot know how incomplete their reference annotations are, making the initial decision of what the NDR threshold should be a mostly uninformed decision. Can bambu approximate how many annotations may be missing based on the NDR of the default model.

### LINKED ISSUE #271: Bambu should tell users when the default model is out performing the fitted model
If users are using a poor annotation set it is possible the fitted model will preform worse due to the increased presence of false negative labels. If the fitted model ends up doing worse than the default model which is trained with a well annotated genome, bambu should let the user know so they can use the default model instead.

### LINKED ISSUE #272: Allow users to train and input new default models
The default model is trained using a nanopore direct rna human dataset. Whilst it does seem broadly applicable, it is possible that in vastly different species and sample preparations, the default model is not suitable. this is mainly important in the niche cases where sample specific training cannot be performed, such as de-novo annotations, or very poor annotations. 
Users should be able to train new models using related by better annotated data and supply it to bambu to use.

### LINKED ISSUE #273: return read count, NDR etc for all transcripts (not just novel transcripts)
read classes which are identical to annotations will be replaced with annotations, read count and other information will be lost. When replacing read classes with annotations, track read count and other metrics (relative read count, subset read count,...) that are specific to the matching read classes.

### PR DESCRIPTION
**Pull request by andredsim**
This pull request involves several features related to the NDR and how it is calculated.

1.  Added the returnModel parameter to isoreParameters. When TRUE the trained model will be appended to the metadata of the rcFile (metadata(se)$model). Users can then pass this model as an isoreParameter to defaultModels when a normal human default model will be preferred. Generally fit should be set to FALSE as well. 

2.  By default the NDR is no longer 0.1 and instead calculates a recommended NDR for the user based on the default 0.1 NDR on human datasets. This is  calculated using the default model and is for cases where the user has poorly annotated genomes and therefore would generally need a more sensitive threshold. 

3. The txScore is now always calculated for the default model and is stored as txScore.noFit. This is used to calculate the suggested NDR threshold and to recommend to users if the default model is likely to be performing better than the fitted model (can occur in cases where the reference annotations are very poor). 

4. All transcripts which have a corresponding read class (and not just novel read classes) will be assigned an NDR instead of only novel transcripts. This means that users could run Bambu with an NDR of 1 to get all possible novel transcripts and filter the NDR manually before the quantification step.

5. The rcFile can be output when quant and discovery are set to false to make accessing a trained model easier.

### CODE REVIEW

File: R/bambu-extendAnnotations-utilityExtend.R [hunk 1]
```diff
@@ -227,6 +282,12 @@ addNewSplicedReadClasses <- function(combinedTranscriptRanges,
     compatibleQhits[!duplicated(compatibleQhits)]] <- "compatible"
   classificationTable$compatible[classificationTable$equal == "equal"] <- ""
   # annotate with transcript and gene Ids
+  equalShits <- subjectHits(ovExon[mcols(ovExon)$equal])
```
Comment (jonathangoeke) [inline R/bambu-extendAnnotations-utilityExtend.R:285]: this line is repeated below

**Review by jonathangoeke — COMMENTED** [review]
see comments

File: DESCRIPTION [hunk 2]
```diff
@@ -85,6 +85,9 @@ Imports:
     Rsamtools,
     methods,
     Rcpp,
-    xgboost
+    xgboost,
+    ROCit,
+    MESS
```
[the remark this answers is not in the document]
Comment (andredsim) [reply DESCRIPTION:91]: replaced with evaluatePerformance 81a05a24559605c3257d316d123b270f072b3074

File: R/bambu.R [hunk 3]
```diff
@@ -130,12 +130,10 @@
 #'     genome = fa.file,  discovery = TRUE, quant = TRUE)
 #' @export
 bambu <- function(reads = NULL, rcFile = NULL, rcOutDir = NULL,
-                  annotations = NULL, genome = NULL, stranded = FALSE, ncore = 1, NDR = 0.1,
-                  yieldSize = NULL, opt.discovery = NULL, opt.em = NULL, trackReads = FALSE, 
-                  returnDistTable = FALSE, discovery = TRUE, quant = TRUE, fusionMode = FALSE, 
-                  verbose = FALSE, lowMemory = FALSE) {
-    if (!(discovery+quant)) stop("At least 1 of discovery and quant must be 
-    TRUE. Rerun with either 1 or both parameters as TRUE")
+    annotations = NULL, genome = NULL, stranded = FALSE, ncore = 1, NDR = 0.1,
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu.R:133]: Yes good catch, I thought was already changed for some reason. 5dc477691b5f1cb9912f753cbc77f42e9e0b6150

File: R/bambu-extendAnnotations-utilityExtend.R [hunk 4]
```diff
@@ -118,6 +127,52 @@ filterTranscriptsByAnnotation <- function(rowDataCombined, annotationGrangesList
   return(extendedAnnotationRanges)
 }
 
+#' calculates an expected NDR based on the annotations'
+recommendNDR <- function(combinedTranscripts, baseline = 0.8, NDR = NULL){
+    message("-- Predicting annotation completeness to determine NDR threshold --")
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-extendAnnotations-utilityExtend.R:132]: line only prints when verbose==TRUE now. 439b62e0096291b914f047952eb6351d9e49a904

File: R/bambu-extendAnnotations-utilityExtend.R (hunk 1 above)
Comment (andredsim) [reply R/bambu-extendAnnotations-utilityExtend.R:285]: Removed the 2nd line, and refactored S[ubject]hits a29d08cd5e50a191fb7568b38ea0b45970530706

File: R/bambu-extendAnnotations-utilityExtend.R [hunk 5]
```diff
@@ -118,6 +127,52 @@ filterTranscriptsByAnnotation <- function(rowDataCombined, annotationGrangesList
...
+    #     combinedTranscripts$maxTxScore[equal])
+    # if(defaultSuitabilityScore < -0.1) {
+    #     message("Unable to accurately predict annotation completeness")
+    #     message("Model agreement score: ", defaultSuitabilityScore)
+    #     #message("Calculated NDR: ", NDR.rec)
+    #     if(is.null(NDR)) message("Therefore, Bambu will use a default NDR of 0.1. To adjust this run Bambu with the NDR parameter")
+    #     message("To calculate annotation completeness in future runs, use a default model trained on more similiar data (see documentation)")
+    #     NDR.rec = 0.1
+    # }
+
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-extendAnnotations-utilityExtend.R:158]: f1ac815273a5cbedb38ee2ad3b182eb425c94af2

File: R/bambu-extendAnnotations-utilityExtend.R [hunk 6]
```diff
@@ -118,6 +127,52 @@ filterTranscriptsByAnnotation <- function(rowDataCombined, annotationGrangesList
...
+    NDRscores = calculateNDR(combinedTranscripts$maxTxScore.noFit, equal)
+    score = combinedTranscripts$maxTxScore.noFit
+    NDR.rec = predict(lm(NDRscores~poly(score,3,raw=TRUE)), newdata=data.frame(score=baseline))
+    NDR.rec = round(NDR.rec,3)
+    if (NDR.rec < 0) NDR.rec = 0
+    message("Calculated NDR: ", NDR.rec)
+    if(NDR.rec > 0.5){
+        message("A high NDR threshold is being recommended by Bambu indicating high levels of novel transcripts")
+        message("We recommend running Bambu with the opt.discovery=list(fitReadClassModel=FALSE)")
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-extendAnnotations-utilityExtend.R:144]: 0251904e7e6283a7b08940268bfbfa6716a5436d

[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-extendAnnotations-utilityExtend.R:144]:
```
        message("A high NDR threshold is being recommended by Bambu indicating high levels of novel transcripts, ",
        "limiting the performance of the trained model")
        message("We recommend training a new model on similiar but well annotated dataset if available ",
        "(https://github.com/GoekeLab/bambu/tree/master#Training-a-model-on-another-speciesdataset-and-applying-it)",
        ", or alternatively running Bambu with opt.discovery=list(fitReadClassModel=FALSE)")

```

File: R/bambu-processReads_scoreReadClasses.R [hunk 7]
```diff
@@ -107,40 +119,109 @@ countPolyATerminals = function(grl, genomeSequence){
...
+    txScoreBaselineSE = predict(lmNDR.SE, newdata=data.frame(NDR.SE=NDR.threshold))
+
+    ROC = NULL
+    ROC.default = NULL
+    ## Compare the trained model AUC to the default model AUC
+    ROC = rocit(score = txScore, class = txFeatures$labels[indexME])
+    ROC$PREC = measureit(score =txScore, class = txFeatures$labels[indexME], measure = c("PREC"))$PREC
+    ROC$PREC[1] = ROC$PREC[2]
+    ROC$PR.AUC = MESS::auc(ROC$TPR, ROC$PREC)
+
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-processReads_scoreReadClasses.R:202]: replaced with evaluatePerformance 81a05a24559605c3257d316d123b270f072b3074

File: R/bambu-extendAnnotations-utilityExtend.R [hunk 8]
```diff
@@ -100,10 +108,11 @@ filterTranscriptsByAnnotation <- function(rowDataCombined, annotationGrangesList
...
     rowDataCombined <- rowDataCombined[filterSet,]
   }
+  #deprecating because it is also done in combineWithAnnotations()
   # remove equals to prevent duplicates when merging with annotations
-  filterSet = rowDataCombined$readClassType != "equalcompatible"
-  exonRangesCombined <- exonRangesCombined[filterSet]
-  rowDataCombined <- rowDataCombined[filterSet,]
+  #filterSet = rowDataCombined$readClassType != "equalcompatible"
+  #exonRangesCombined <- exonRangesCombined[filterSet]
+  #rowDataCombined <- rowDataCombined[filterSet,]
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-extendAnnotations-utilityExtend.R:115]: ca00daf95e5eb5370712659bdc65155e98a876e3

File: R/bambu-processReads_scoreReadClasses.R [hunk 9]
```diff
@@ -107,40 +119,109 @@ countPolyATerminals = function(grl, genomeSequence){
...
+    txScore.default = predict(defaultModels$transcriptModelME, as.matrix(features))[indexME] 
+    ROC.default = rocit(score = txScore.default, class = txFeatures$labels[indexME])
+    ROC.default$PREC = measureit(score =txScore.default, class = txFeatures$labels[indexME], measure = c("PREC"))$PREC
+    ROC.default$PREC[1] = ROC.default$PREC[2]
+    ROC.default$PR.AUC = MESS::auc(ROC.default$TPR, ROC.default$PREC)
+    if(verbose){
+        message("On the dataset the new trained model achieves a ROC AUC of ",
+            signif(ROC$AUC,3),  " and a Precision-Recall AUC of ", signif(ROC$PR.AUC,3), ".", 
+            "This is compared to the original default model which achiveves a ROC AUC of ",
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-processReads_scoreReadClasses.R:211]: ca00daf95e5eb5370712659bdc65155e98a876e3

File: R/bambu-processReads_scoreReadClasses.R [hunk 10]
```diff
@@ -107,40 +119,109 @@ countPolyATerminals = function(grl, genomeSequence){
 
 
 #' calculates a score based on how likely a read class is full length
-getTranscriptScore = function(rowData, defaultModels, nrounds = 50, fit = TRUE){
+getTranscriptScore = function(rowData, model = NULL, defaultModels, nrounds = 50, fit = TRUE){
     txFeatures = prepareTranscriptModelFeatures(rowData)
     features = dplyr::select(txFeatures,!c(labels))
-    if(checkFeatures(txFeatures) & fit){
+    if(fit & !is.null(model)){
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-processReads_scoreReadClasses.R:125]: ca00daf95e5eb5370712659bdc65155e98a876e3

File: R/bambu-processReads_scoreReadClasses.R [hunk 11]
```diff
@@ -107,40 +119,109 @@ countPolyATerminals = function(grl, genomeSequence){
 
 
 #' calculates a score based on how likely a read class is full length
-getTranscriptScore = function(rowData, defaultModels, nrounds = 50, fit = TRUE){
+getTranscriptScore = function(rowData, model = NULL, defaultModels, nrounds = 50, fit = TRUE){
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-processReads_scoreReadClasses.R:122]: ca00daf95e5eb5370712659bdc65155e98a876e3

File: tests/testthat/test_txRange.R [hunk 12]
```diff
@@ -73,7 +73,7 @@ test_that("getTranscriptScore() calculates the correct score", {
                                    package = "bambu"))
   se=seExpected
   thresholdIndex = which(rowData(se)$readCount >= 2)
-  txScore = getTranscriptScore(rowData(se)[thresholdIndex,], 
+  txScore = getTranscriptScore(rowData(se)[thresholdIndex,], NULL,
```
[the remark this answers is not in the document]
Comment (andredsim) [reply tests/testthat/test_txRange.R:76]: b4734e830af9cdaacf1cc4dd39355094cfa628d1

File: R/globals.R [hunk 13]
```diff
@@ -41,7 +41,7 @@ if (getRversion() >= "2.15.1") {
         "NSampleTxScore.combined", "NSampleTxScore.new", "maxTxScore.combined",
         "maxTxScore.new" ,"txScore" ,"row_id" ,"txScore", "sample_id", 
         "sample_name", "txScore", "readCount_tmp", "group", "row_id", "rowID", 
-        "sumReadCount", "maxTxScore", "NSampleReadCount", "NSampleReadProp", 
+        "sumReadCount", "maxTxScore.noFit", "maxTxScore", "NSampleReadCount", "NSampleReadProp",
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/globals.R:44]: 05f4cf209917a8d8b14d65c496023e89b8e40e3c

File: R/bambu-processReads_scoreReadClasses.R [hunk 14]
```diff
@@ -31,10 +32,21 @@ scoreReadClasses = function(se, genomeSequence, annotations, defaultModels,
     rowData(se)[names(newRowData)] = NA
     rowData(se)[thresholdIndex,names(newRowData)] = newRowData
     
-    txScore = getTranscriptScore(rowData(se)[thresholdIndex,], 
+    model = NULL
+    if (fit) model = trainBambu(se, verbose = verbose)
+    if(returnModel) metadata(se)$model = model
+    txScore = getTranscriptScore(rowData(se)[thresholdIndex,], model,
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-processReads_scoreReadClasses.R:38]: 0f911db925b43a598cb46c260f86b9d33ebf84dd

File: R/bambu-processReads_scoreReadClasses.R [hunk 15]
```diff
@@ -107,40 +119,109 @@ countPolyATerminals = function(grl, genomeSequence){
...
+        }
     }
     txScore[which(rowData$numExons==1)] =
         txScoreSE[which(rowData$numExons==1)]
     return(txScore)
 }
 
+#' Function to train a model for use on other data
+#' @noRd
+trainBambu <- function(rcFile = NULL, min.readCount = 2, nrounds = 50, NDR.threshold = 0.1, verbose = TRUE) {
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-processReads_scoreReadClasses.R:160]: Yes this function is accessed directly so I added documentation
6d44e1aa8b49f27890f19226f04e892cc788d0dc

File: R/bambu-processReads_scoreReadClasses.R [hunk 16]
```diff
@@ -107,40 +119,109 @@ countPolyATerminals = function(grl, genomeSequence){
...
+    ROC.default$PREC[1] = ROC.default$PREC[2]
+    ROC.default$PR.AUC = MESS::auc(ROC.default$TPR, ROC.default$PREC)
+    if(verbose){
+        message("On the dataset the new trained model achieves a ROC AUC of ",
+            signif(ROC$AUC,3),  " and a Precision-Recall AUC of ", signif(ROC$PR.AUC,3), ".", 
+            "This is compared to the original default model which achiveves a ROC AUC of ",
+            signif(ROC.default$AUC,3), " and a Precision-Recall AUC of ", signif(ROC.default$PR.AUC,3))
+    }
+
+    return(list(transcriptModelME = transcriptModelME,
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-processReads_scoreReadClasses.R:215]: 56ae92847edebc9e190d10d9a6900a82595ed45f

File: R/bambu-processReads_scoreReadClasses.R [hunk 17]
```diff
@@ -107,40 +119,109 @@ countPolyATerminals = function(grl, genomeSequence){
...
+            labels.train=txFeatures$labels[indexSE], 
+            nrounds = nrounds, show.cv=FALSE)
+        txScoreSE = predict(transcriptModelSE, as.matrix(features))[indexSE]
+    }
+    ##Calculate the txScore baseline
+    NDR = calculateNDR(txScore, txFeatures$labels[indexME])
+    NDR.SE = calculateNDR(txScoreSE, txFeatures$labels[indexSE])
+    #lm of NDR vs txScore
+    lmNDR = lm(txScore~poly(NDR,3,raw=TRUE))
+    txScoreBaseline = predict(lmNDR, newdata=data.frame(NDR=NDR.threshold))
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-processReads_scoreReadClasses.R:191]: 56ae92847edebc9e190d10d9a6900a82595ed45f

File: R/bambu_utilityFunctions.R [hunk 18]
```diff
@@ -31,8 +31,14 @@ setIsoreParameters <- function(isoreParameters){
         min.txScore.multiExon = 0,
         min.txScore.singleExon = 1,
         fitReadClassModel = TRUE,
+        defaultModels = defaultModels,
+        returnModel = FALSE,
+        txScoreBaseline = defaultModels$txScoreBaseline,
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu_utilityFunctions.R:36]: 394daee0ad3c2aeaa6fe1bf859e442e59c8ca9c4
replaced by baselineFDR

File: R/bambu_utilityFunctions.R [hunk 19]
```diff
@@ -31,8 +31,14 @@ setIsoreParameters <- function(isoreParameters){
...
         min.readFractionByEqClass = 0,
         prefix = "") 
+    if("defaultModels" %in% names(isoreParameters) & !("txScoreBaseline" %in% names(isoreParameters))){
+        isoreParameters["txScoreBaseline"]=isoreParameters$defaultModels$txScoreBaseline
+    }
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu_utilityFunctions.R:41]: removed in 394daee0ad3c2aeaa6fe1bf859e442e59c8ca9c4

File: R/bambu-extendAnnotations.R [hunk 20]
```diff
@@ -33,6 +33,7 @@ bambu.extendAnnotations <- function(readClassList, annotations, NDR,
         isoreParameters[['min.readFractionByEqClass']],
         fusionMode = fusionMode,
         prefix = isoreParameters[["prefix"]],
+        txScoreBaseline = isoreParameters[["txScoreBaseline"]],
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-extendAnnotations.R:36]: 394daee0ad3c2aeaa6fe1bf859e442e59c8ca9c4
now uses baselineFDR

File: R/bambu-extendAnnotations-utilityExtend.R [hunk 21]
```diff
@@ -118,6 +127,52 @@ filterTranscriptsByAnnotation <- function(rowDataCombined, annotationGrangesList
   return(extendedAnnotationRanges)
 }
 
+#' calculates an expected NDR based on the annotations'
+recommendNDR <- function(combinedTranscripts, baseline = 0.8, NDR = NULL){
```
[the remark this answers is not in the document]
Comment (andredsim) [reply R/bambu-extendAnnotations-utilityExtend.R:131]: 394daee0ad3c2aeaa6fe1bf859e442e59c8ca9c4
argument now has baselineFDR and NDR

**Pull request merged**